src/iterator.jl

```diff
@@ -0,0 +1,28 @@
+struct CuIterator{B,P}
+    batches::B
+    pool::Vector{P}
+    pool_limit::Int
+end
+
+CuIterator(batches, pool=Any[]) = CuIterator(batches, pool, 10)
+
+pool_key(x) = eltype(x) => size(x)
+
+function Base.iterate(c::CuIterator, state...)
+    item = iterate(c.batches, state...)
+    if item === nothing
```

Comment:
Note that if we take this kind of approach for freeing the pool, callers which break out of iteration early won't get the nice automatic freeing behavior